File: Bookcase core/src/main/java/cz/vhromada/book/domain/Category.java
package cz.vhromada.book.domain;

import java.util.Objects;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.SequenceGenerator;
import javax.persistence.Table;

import cz.vhromada.common.Movable;

/**
 * A class represents category.
 *
 * @author Vladimir Hromada
 */
@Entity
@Table(name = "categories")
public class Category implements Movable {

    /**
     * SerialVersionUID
     */
    private static final long serialVersionUID = 1L;

    /**
     * ID
     */
    @Id
    @SequenceGenerator(name = "category_generator", sequenceName = "categories_sq", allocationSize = 0)
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "category_generator")
    private Integer id;

    /**
     * Name
     */
    @Column(name = "category_name")
    private String name;

    /**
     * Position
     */
    private Integer position;

    @Override
    public Integer getId() {
        return id;
    }

    @Override
    public void setId(final Integer id) {
        this.id = id;
    }

    /**
     * Returns name.
     *
     * @return name
     */
    public String getName() {
        return name;
    }

    /**
     * Sets a new value to name.
     *
     * @param name new value
     */
    public void setName(final String name) {
        this.name = name;
    }

    @Override
    public Integer getPosition() {
        return position;
    }

    @Override
    public void setPosition(final Integer position) {
        this.position = position;
    }

    @Override
    public boolean equals(final Object obj) {
        if (this == obj) {
            return true;
        }

        if (!(obj instanceof Category) || id == null) {
            return false;
        }

        return id.equals(((Category) obj).id);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id);
    }

    @Override
    public String toString() {
        return String.format("Category [id=%d, name=%s, position=%d]", id, name, position);
    }

}

File: Bookcase core/src/main/kotlin/cz/vhromada/book/domain/Category.kt
package cz.vhromada.book.domain

import cz.vhromada.book.common.Movable
import java.util.Objects
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.GeneratedValue
import javax.persistence.GenerationType
import javax.persistence.Id
import javax.persistence.SequenceGenerator
import javax.persistence.Table

/**
 * A class represents category.
 *
 * @author Vladimir Hromada
 */
@Entity
@Table(name = "categories")
data class Category(

    @field:Id
    @field:SequenceGenerator(name = "category_generator", sequenceName = "categories_sq", allocationSize = 0)
    @field:GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "category_generator")
    override var id: Int?,

    /**
     * Name
     */
    @field:Column(name = "category_name")
    val name: String,

    override var position: Int

) : Movable {

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }

        return if (other !is Category || id == null) {
            false
        } else {
            id == other.id
        }
    }

    override fun hashCode(): Int {
        return Objects.hashCode(id)
    }

}

